[Order Creation] Prevent merchant from adding parent product to an order via barcode scanning

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12566
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Scanning a barcode of a parent product previously added it to the order.

This isn’t possible in other ways, as normally a variation should be selected.

The implementation here shows the new order with a notice informing the user of the reason their product isn’t added. This matches Android.

A future improvement might be to open the product variation picker for the relevant product.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Launch the app, and log in to a store with variable products that have barcodes set for their SKU. Ensure the parent product also has a barcode SKU
2. Tap `Orders`
3. Tap scan
4. Scan the parent product barcode
5. Observe that the order form opens with a notice to say why the parent product wasn't added
6. Retry with a variation barcode
7. Observe that the order form opens with the variation
8. Try with an ordinary product, verify that it's added as expected

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/2472348/65c84b69-d5e7-4a6f-9a35-00351b8a8977



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
12566 Prevent adding parent product via barcode

Scanning a barcode of a parent product previously added it to the order.

This isn’t possible in other ways, as normally a variation should be selected.

The implementation here shows the new order with a notice informing the user of the reason their product isn’t added. This matches Android.

A future improvement might be to open the product variation picker for the relevant product.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
12566 Add barcode scan fix to 18.5 release notes

joshheald committed:
12566 Apply barcode fix to button on Order Form

Parent variable products should not be able to be added to the order when done from the Order List barcode scanner, or the Order Form barcode scanner.

Moving this code ensures it covers both routes.

joshheald committed:
Merge branch 'trunk' into bug/12566-prevent-adding-parent-product-to-an-order

iamgabrielma reviewed (approved):
Works nicely when added parent variable products from either the order list, or the editable order. Adds other products and variations normally.

LGTM :shipit:
